finite_difference/src/field_mod.f90

```diff
@@ -1064,14 +1214,55 @@ subroutine halo_exchange(self, depth)
     ! Locals
     integer :: exch  !> Handle for exchange
 
-    ! Synchronize the data if the data is in an external device
-    self%data = self%get_data()
+    ! Make sure the data from each halo is in the host before the communication
+    call self%read_halo_from_device(comm=JPlus, blocking=.false.)
```

Comment:
This reminds me of the discussions we're having with Mellanox/NVIDIA. We could (I think) interleave the halo exchange calls with these copies by only setting `comm1` in each call to `exchange_generic`. Anyway, that's not something for this PR although if you could add a TODO that would be good.

Reply:
I added TODO #58 to capture the interleaving of device and mpi comms, but also if there is external useful computation that could be executed while the communication is ongoing.